lib/spack/spack/package.py

```diff
@@ -1065,11 +1072,20 @@ def do_stage(self, mirror_only=False):
         if not self.spec.concrete:
             raise ValueError("Can only stage concrete packages.")
 
-        self.do_fetch(mirror_only)     # this will create the stage
-        self.stage.expand_archive()
+        # Always create the stage directory at this point.  Why?  A no-code
+        # package may want to use the installation process to install metadata.
+        self.stage.create()
+
+        # Fetch/expand any associated code.
+        if self.has_code:
+            self.do_fetch(mirror_only)
+            self.stage.expand_archive()
 
-        if not os.listdir(self.stage.path):
-            raise FetchError("Archive was empty for %s" % self.name)
+            if not os.listdir(self.stage.path):
+                raise FetchError("Archive was empty for %s" % self.name)
+        else:
+            # Support for post-install hooks requires a stage.source_path
+            mkdirp(self.stage.source_path)
```

Comment:
It's code so yes, it could, though the directory's creation is conceptually part of fetch-and-expand, not basic directory creation.  From a conceptual point of view (or symmetry), the current location is more appropriate.

What is the issue with having the directory being created here?

Reply:
> What is the issue with having the directory being created here?

If instantiating the `Stage` doesn't guarantee that the `source_path` exists, it must be created by all users. So things are done correctly here, but the user has to remember to do this everywhere they want to use the `Stage`.

Another option is that the `Stage.source_path` property could instantiate the directory at the time that it is accessed.

Reply:
I'm not familiar with the former use case.  Why would `do_stage` not always be called as part of the install process?

As for the latter, the point of adding `source_path` was to make it available to packages *before* staging and, IIRC, @tgamblin was very clear that the `source-path` should only be created during staging.

Reply:
I don't think my comment conflicts with

> As for the latter, the point of adding source_path was to make it available to packages before staging

but it does conflict with

> tgamblin was very clear that the source-path should only be created during staging.

I forgot that. We can leave this as is. 

> I'm not familiar with the former use case. Why would do_stage not always be called as part of the install process?

It should be called as part of the install process. However, folding the creation of `.source_path` into a `Stage` function might simplify its use elsewhere (e.g. outside of `Package`).

Reply:
If such a change is made, it is important to keep in mind that `Stage` assumes that if`source_path` exists, then the archive has been expanded (i.e., staging was successful).